fclient/performjoin.go

```diff
@@ -0,0 +1,238 @@
+	// contain signatures that we don't know about.
+	if len(respSendJoin.Event) > 0 {
+		var remoteEvent *gomatrixserverlib.Event
+		remoteEvent, err = verImpl.NewEventFromUntrustedJSON(respSendJoin.Event)
+		if err == nil && isWellFormedMembershipEvent(
+			remoteEvent, input.RoomID, input.UserID,
+		) {
+			event = remoteEvent
+		}
+	}
+
+	// Sanity-check the join response to ensure that it has a create
+	// event, that the room version is known, etc.
+	authEvents := respSendJoin.AuthEvents.UntrustedEvents(respMakeJoin.RoomVersion)
+	if err = checkEventsContainCreateEvent(authEvents); err != nil {
+		return nil, nil, fmt.Errorf("sanityCheckAuthChain: %w", err)
+	}
+
+	// Process the join response in a goroutine. The idea here is
+	// that we'll try and wait for as long as possible for the work
+	// to complete, but if the client does give up waiting, we'll
+	// still continue to process the join anyway so that we don't
+	// waste the effort.
+	// TODO: Can we expand Check here to return a list of missing auth
+	// events rather than failing one at a time?
+	respState, err = gomatrixserverlib.CheckSendJoinResponse(
+		context.Background(),
+		respMakeJoin.RoomVersion, &respSendJoin,
+		input.KeyRing,
+		event,
+		input.EventProvider(ctx, fedClient, input.KeyRing, origin, input.ServerName),
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("respSendJoin.Check: %w", err)
+	}
+
+	// If we successfully performed a send_join above then the other
+	// server now thinks we're a part of the room. Send the newly
+	// returned state to the roomserver to update our local view.
+	if input.Unsigned != nil {
+		event, err = event.SetUnsigned(input.Unsigned)
+		if err != nil {
+			// non-fatal, log and continue
+			logrus.WithError(err).Errorf("Failed to set unsigned content")
+		}
+	}
+
+	return event.Headered(respMakeJoin.RoomVersion), respState, nil
+}
+
+func setDefaultRoomVersionFromJoinEvent(
+	joinEvent gomatrixserverlib.EventBuilder,
+) gomatrixserverlib.RoomVersion {
+	// if auth events are not event references we know it must be v3+
+	// we have to do these shenanigans to satisfy sytest, specifically for:
+	// "Outbound federation rejects m.room.create events with an unknown room version"
+	hasEventRefs := true
+	authEvents, ok := joinEvent.AuthEvents.([]interface{})
+	if ok {
+		if len(authEvents) > 0 {
+			_, ok = authEvents[0].(string)
+			if ok {
+				// event refs are objects, not strings, so we know we must be dealing with a v3+ room.
+				hasEventRefs = false
+			}
+		}
+	}
+
+	if hasEventRefs {
+		return gomatrixserverlib.RoomVersionV1
+	}
+	return gomatrixserverlib.RoomVersionV4
+}
+
+// isWellFormedMembershipEvent returns true if the event looks like a legitimate
+// membership event.
+func isWellFormedMembershipEvent(event *gomatrixserverlib.Event, roomID string, userID *spec.UserID) bool {
+	if membership, err := event.Membership(); err != nil {
+		return false
+	} else if membership != spec.Join {
+		return false
```

Comment:
This is misleading. The function is `isWellFormedMembershipEvent` _not_ `isWellFormedJoinMemberEvent` so failing the check if the membership != join is deceptive. Please rename this function.